routing/router.go

```diff
@@ -1666,119 +1712,337 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
+		// In case of success we atomically store the db payment and
+		// move the payment to the success state.
+		err = r.cfg.Control.Success(payment.PaymentHash, result.Preimage)
+		if err != nil {
+			log.Errorf("Unable to succeed payment "+
+				"attempt: %v", err)
+			return [32]byte{}, nil, err
 		}
 
-		lastError = err
+		// Terminal state, return the preimage and the route
+		// taken.
+		return result.Preimage, &attempt.Route, nil
 	}
-}
 
-// sendPaymentAttempt tries to send the payment via the specified route. If
-// successful, it returns the obtained preimage. If an error occurs, the last
-// bool parameter indicates whether this is a final outcome or more attempts
-// should be made.
-func (r *ChannelRouter) sendPaymentAttempt(paySession *paymentSession,
-	route *route.Route, paymentHash [32]byte) ([32]byte, bool, error) {
+}
 
-	log.Tracef("Attempting to send payment %x, using route: %v",
-		paymentHash, newLogClosure(func() string {
-			return spew.Sdump(route)
-		}),
-	)
+// parseFailedPayment determines the appropriate failure message to return to
+// a user initiated payment. The three cases handled are:
+// 1) A local failure, which should already plaintext.
+// 2) A resolution from the chain arbitrator,
+// 3) A failure from the remote party, which will need to be decrypted using the
+//      payment deobfuscator.
+func (r *ChannelRouter) parseFailedPayment(paymentHash lntypes.Hash,
+	rt *route.Route, sessionKey *btcec.PrivateKey, selfKey *btcec.PublicKey,
+	result *htlcswitch.PaymentResult) (*htlcswitch.ForwardingError, error) {
 
-	preimage, err := r.sendToSwitch(route, paymentHash)
-	if err == nil {
-		return preimage, true, nil
-	}
+	var failure *htlcswitch.ForwardingError
 
-	log.Errorf("Attempt to send payment %x failed: %v",
-		paymentHash, err)
+	switch {
 
-	finalOutcome := r.processSendError(paySession, route, err)
+	// The payment never cleared the link, so we don't need to decrypt the
+	// error, simply decode it them report back to the user.
+	case result.Type == htlcswitch.PaymentResultLocalError:
+		var userErr string
+		rd := bytes.NewReader(result.Reason)
+		failureMsg, err := lnwire.DecodeFailure(rd, 0)
+		if err != nil {
+			userErr = fmt.Sprintf("unable to decode onion failure, "+
+				"htlc with hash(%v): %v",
+				paymentHash, err)
+			log.Error(userErr)
+
+			// As this didn't even clear the link, we don't need to
+			// apply an update here since it goes directly to the
+			// router.
+			failureMsg = lnwire.NewTemporaryChannelFailure(nil)
+		}
+		failure = &htlcswitch.ForwardingError{
+			ErrorSource:    selfKey,
+			ExtraMsg:       userErr,
+			FailureMessage: failureMsg,
+		}
 
-	return [32]byte{}, finalOutcome, err
-}
+	// A payment had to be timed out on chain before it got past the first
+	// hop. In this case, we'll report a permanent channel failure as this
+	// means us, or the remote party had to go on chain.
+	case result.Type == htlcswitch.PaymentResultResolutionError && result.Reason == nil:
```

Comment:
It looks like there's a case where it can be non-nil: https://github.com/lightningnetwork/lnd/blob/a8fa4094ba61e26bd9855e9743883e3a234d383a/htlcswitch/switch.go#L1184

Made it more clear by doing it this way: https://github.com/lightningnetwork/lnd/pull/3063/commits/77ffe6c692f475dd0bfa89c9fd7530ceecae7b97

Reply:
In that snippet, what does actually happen? We encrypt a message to ourselves? That is probably not the case, so then it must be an intermediate hop. When the response arrrives at the sender hop, `isResolution` is no longer true?

Reply:
Not sure tbh, I just wanted to keep the current behavior in case the `reason == nil` check is there for a reason.

Maybe @cfromknecht can shed some light.